### PR TITLE
allow using numpy dtypes for symbol and ndarray functions

### PR DESCRIPTION
**Pull request by piiswrong**
@sxjscience

### CODE REVIEW

File: python/mxnet/_ctypes/ndarray.py [hunk 1]
```diff
@@ -99,8 +100,11 @@ def generic_ndarray_function(*args, **kwargs):
             else:
                 pos_args.append(str(i))
 
-        assert len(pos_args) <= len(arguments)
+        if len(pos_args) > len(arguments)
+            raise ValueError("Too many positional arguments")
         kwargs.update(zip(arguments[:len(pos_args)], pos_args))
+        if 'dtype' in kwargs:
+            kwargs['dtype'] = np.dtype(kwargs['dtype']).name
```
Comment (sxjscience) [inline python/mxnet/_ctypes/ndarray.py]: This will force the users to use types from numpy.

Comment (piiswrong) [inline python/mxnet/_ctypes/ndarray.py]: no. np.dtype('float32') works

**Comment by sxjscience** [comment]
I think this PR is good. I can remove the python wrapper of arange in ndarray.py now. However, I still have some problems directly using `mx.sym.arange`. Sometimes I want it to be
```python
a = mx.sym.arange(10, dtype=float)
```
which should be equivalent to 
```python
a = mx.nd.arange(10, dtype=float)
```
But there are checks https://github.com/dmlc/mxnet/blob/nnvm/python/mxnet/cython/symbol.pyx#L137-L139, https://github.com/dmlc/mxnet/blob/nnvm/python/mxnet/_ctypes/symbol.py#L60-L65 that disable this behavior.

**Comment by piiswrong** [comment]
I don't think dtype=float should be allowed. Its ambiguous. Should it mean float32 or 64?

**Comment by sxjscience** [comment]
@piiswrong The problem is that the current API do not allow missing keys when setting arguments.
The following is allowed in the current setting:
```python
a = mx.sym.arange(start=10, dtype=np.float32)
```
The following is not allowed:
```python
a = mx.sym.arange(10, dtype=np.float32)
```

**Comment by sxjscience** [comment]
@piiswrong Also I find that the logic in ctypes and cython seems to be different.
In ctypes (non-symbol value in args/kwargs is not allowed): 
```python
        for arg in args:
            if not isinstance(arg, SymbolBase):
                raise TypeError('Compose expect `Symbol` as arguments')
        for val in kwargs.values():
            if not isinstance(val, SymbolBase):
                raise TypeError('Compose expect `Symbol` as arguments')
```
In cython (non-symbol value in args is not allowed):
```python
        if key_vargs:
            if key_vargs not in kwargs:
                sparam_keys.push_back(c_str(key_vargs))
                sparam_vals.push_back(c_str(str(len(args))))

        if len(kwargs) != 0:
            for k, v in kwargs.items():
                if isinstance(v, SymbolBase):
                    ssymbol_keys.push_back(c_str(k))
                    symbol_args.push_back((<SymbolBase>v).chandle)
                else:
                    sparam_keys.push_back(c_str(k))
                    sparam_vals.push_back(c_str(str(v)))

        if len(args) != 0:
            if symbol_args.size() != 0:
                raise TypeError("compose only accept input Symbols\
                    either as positional or keyword arguments, not both")
            for v in args:
                if not isinstance(v, SymbolBase):
                    raise TypeError('Compose expect `Symbol` as arguments')
                symbol_args.push_back((<SymbolBase>v).chandle)

**Comment by piiswrong** [comment]
@tqchen Could you have a look?

**Comment by tqchen** [comment]
Current API in general do not support positional argument that is non symbolic

**Comment by piiswrong** [comment]
ndarray api actually supports it.

**Comment by tqchen** [comment]
I think it is a good PR. But we could keep some of the front-end wrapping when necessarily to solve the mixed arg and kwarg cases. If for imperative api it is already supported, we can remove those

**Pull request merged**